### PR TITLE
Decision dag: union the set of values when there are two predecessor states.

### PR DESCRIPTION
**Pull request by gafter**
If a state has no set of values for a temp on one of the edges, that means it is unrestricted.
This fixes a bug in decision dag construction.  I was not able to construct a test that is
affected, but the change makes a visible difference in an internal dump of the decision dag
for the newly added test `MultiplePathsToState_01`, which we walked through in a compiler
team meeting on 2020-09-09.

### CODE REVIEW

**Comment by gafter** [comment]
May I please have a couple of reviews of this from the compiler team?

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs [hunk 1]
```diff
@@ -679,32 +681,23 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                 var state = new DagState(cases, remainingValues);
                 if (uniqueState.TryGetValue(state, out DagState? existingState))
                 {
-                    var newRemainingValues = existingState.RemainingValues;
-                    bool changed = false;
+                    // We found an existing state that matches.  Update its set of possible remaining values
+                    // of each temp by taking the union of the sets on each incoming edge.
+                    var newRemainingValues = ImmutableDictionary<BoundDagTemp, IValueSet>.Empty;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:686]: Could we use a `Builder` instead? I'd imagine that constructing immutable dictionaries incrementally with `SetItem` would create more allocations :-/

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1), with suggestion to use a builder

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs:686]: It turns out that these dictionaries are typically very small (usually zero or one item). I don't know which would overall be faster, but a builder is certainly an option. I'm going to leave it this way in this PR though.

---
In reply to: [486774228](https://github.com/dotnet/roslyn/pull/47567#discussion_r486774228) [](ancestors = 486774228)

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs (hunk 1 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs]: Actually, I took your suggestion and used a builder.

---
In reply to: [487215991](https://github.com/dotnet/roslyn/pull/47567#discussion_r487215991) [](ancestors = 487215991,486774228)

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Comment by jaredpar** [comment]
The Spanish previous failure is "stuck" on this PR. I've verified it passed on this run and integration tests previously passed. Forcing the merge.

**Pull request merged**